tianshou/data/batch.py

```diff
@@ -691,26 +798,53 @@ def __len__(self) -> int:
         """Return len(self)."""
         r = []
         for v in self.__dict__.values():
-            if isinstance(v, Batch) and v.is_empty():
+            if isinstance(v, Batch) and v.is_empty(recurse=True):
                 continue
             elif hasattr(v, '__len__') and (not isinstance(
                     v, (np.ndarray, torch.Tensor)) or v.ndim > 0):
                 r.append(len(v))
             else:
-                raise TypeError("Object of type 'Batch' has no len()")
+                raise TypeError(f"Object {v} in {self} has no len()")
         if len(r) == 0:
-            raise TypeError("Object of type 'Batch' has no len()")
+            raise TypeError(f"Object {self} has no len()")
         return min(r)
 
-    def is_empty(self):
-        return not any(
-            not x.is_empty() if isinstance(x, Batch)
-            else hasattr(x, '__len__') and len(x) > 0 for x in self.values())
+    def is_empty(self, recurse: bool = False):
+        """
+        Test if a Batch is empty. If ``recurse=True``, it further tests the
+        values of the object; else it only tests the existence of any key.
+
+        ``b.is_empty(recurse=True)`` is mainly used to distinguish
+        ``Batch(a=Batch(a=Batch()))`` and ``Batch(a=1)``. They both raise
+        exceptions when applied to ``len()``, but the former can be used in
+        ``cat``, while the latter is a scalar and cannot be used in ``cat``.
+
+        Another usage is in ``__len__``, where we have to skip checking the
+        length of recursely empty Batch.
+        ::
+
+        >>>Batch().is_empty()
+        True
+        >>>Batch(a=Batch(), b=Batch(c=Batch())).is_empty()
+        False
+        >>>Batch(a=Batch(), b=Batch(c=Batch())).is_empty(recurse=True)
+        True
+        >>>Batch(d=1).is_empty()
+        False
+        >>>Batch(a=np.float64(1.0)).is_empty()
+        False
```

Comment:
Should add 4 space before `>>>` and another space after `>>>`:
```python
    >>> Batch().is_empty()
    True
```
I think it is good to render the documentation and take a look before the commit.

Reply:
Maybe we can add the doc render as a test, too, just like pep8 check.

Reply:
Agree